server/leafnode.go

```diff
@@ -0,0 +1,1074 @@
+	cinfo := leafConnectInfo{
+		TLS:  tlsRequired,
+		Name: c.srv.info.ID,
+	}
+
+	// Check for credentials first, that will take precedence..
+	if creds := c.leaf.remote.Credentials; creds != "" {
+		c.Debugf("Authenticating with credentials file %q", c.leaf.remote.Credentials)
+		contents, err := ioutil.ReadFile(creds)
+		if err != nil {
+			c.Errorf("%v", err)
+			return
+		}
+		defer wipeSlice(contents)
+		items := credsRe.FindAllSubmatch(contents, -1)
+		if len(items) < 2 {
+			c.Errorf("Credentials file malformed")
+			return
+		}
+		// First result should be the user JWT.
+		// We copy here so that the file containing the seed will be wiped appropriately.
+		raw := items[0][1]
+		tmp := make([]byte, len(raw))
+		copy(tmp, raw)
+		// Seed is second item.
+		kp, err := nkeys.FromSeed(items[1][1])
+		if err != nil {
+			c.Errorf("Credentials file has malformed seed")
+			return
+		}
+		// Wipe our key on exit.
+		defer kp.Wipe()
+
+		sigraw, _ := kp.Sign(c.nonce)
+		sig := base64.RawURLEncoding.EncodeToString(sigraw)
+		cinfo.JWT = string(tmp)
+		cinfo.Sig = sig
+	} else if userInfo := c.leaf.remote.URL.User; userInfo != nil {
+		cinfo.User = userInfo.Username()
+		pass, _ := userInfo.Password()
+		cinfo.Pass = pass
+	}
+
+	b, err := json.Marshal(cinfo)
+	if err != nil {
+		c.Errorf("Error marshaling CONNECT to route: %v\n", err)
+		c.closeConnection(ProtocolViolation)
+		return
+	}
+	c.sendProto([]byte(fmt.Sprintf(ConProto, b)), true)
+}
+
+// Called when an inbound leafnode connection is either accepted.
+func (s *Server) createLeafNode(conn net.Conn, remote *RemoteLeafOpts) *client {
+	// Snapshot server options.
+	opts := s.getOpts()
+
+	maxPay := int32(opts.MaxPayload)
+	maxSubs := int32(opts.MaxSubs)
+	// For system, maxSubs of 0 means unlimited, so re-adjust here.
+	if maxSubs == 0 {
+		maxSubs = -1
+	}
+	now := time.Now()
+
+	c := &client{srv: s, nc: conn, kind: LEAF, opts: defaultOpts, mpay: maxPay, msubs: maxSubs, start: now, last: now}
+	c.leaf = &leaf{smap: map[string]int32{}}
+
+	// Determines if we are soliciting the connection or not.
+	var solicited bool
+
+	if remote != nil {
+		solicited = true
+		// Users can bind to any local account, if its empty
+		// we will assume the $G account.
+		if remote.LocalAccount == "" {
+			remote.LocalAccount = globalAccountName
+		}
+		// FIXME(dlc) - Make this resolve at startup.
+		c.acc, _ = s.LookupAccount(remote.LocalAccount)
+		// Make sure we register with the account here.
+		c.registerWithAccount(c.acc)
+		c.leaf.remote = remote
+	}
+
+	// Grab server variables
+	s.mu.Lock()
+	info := s.leafNodeInfo
+	s.mu.Unlock()
+
+	// Grab lock
+	c.mu.Lock()
+
+	c.initClient()
+
+	c.Debugf("LeafNode connection created")
+
+	if solicited {
+		// We need to wait here for the info, but not for too long.
+		b := make([]byte, MAX_CONTROL_LINE_SIZE)
+		c.nc.SetReadDeadline(time.Now().Add(time.Second))
+		n, err := c.nc.Read(b)
```

Comment:
You can't do that. You are not guaranteed at all to get the whole protocol in one read (even if underlying buffer is big enough to hold it). You need either a loop here until it is complete, or maybe use buffered io reader where you could use ReadLine(), or make the processing completely async by starting the read loop and sending the connect only when processing the INFO there. This is what I do with gateways. Check `processGatewayInfo()`.

Reply:
I thought about buffered but problem is if you read more than you want underneath. You would have to still handle it via the parse method in some form or fashion. Will fix though.

Reply:
Actually you are right, b/c spinning up go routine and processing async works but not when we have TLS.